### ISSUE TITLE
UCT: need way to avoid packing/unpacking the registration key

### ISSUE THREAD

**Issue by hjelmn** [openucx/ucx#2408]
This is a feature request.

It seems that in my btl/ucx prototype there is a small but noticeable percentage of time spent unpacking the registration key. It would be better if there was a way to get a uct_rkey_t  from a memory handle so that I can pass that around and not a packed key. Would also need a way to get the uct_rkey_t size from the memory domain.

**Comment by yosefe** [comment]
do you have key pack/unpack in data path? or the intention is to optimize window setup?
it's not possible to pass around uct_rkey_t, since it will not be valid in the context of the remote process, that is what pack/unpack is for..
do you know what takes so much time in pack/unpack? which MD (transport) is it?

**Comment by hjelmn** [comment]
It is in the data path. This is with the ugni memory domain. It adds about 6.4% overhead (according to vtune) to the small message bandwidth. More than half the 6.4% is malloc (3.9%). vtune is not giving anything more helpful than that.

If rkey unpack can be optimized heavily that would be great. Was kinda hoping I could get direct access to the 16-byte key on ugni or the 4-byte (or whatever it is now) on ib.

**Comment by yosefe** [comment]
rkey unpack is not intended for data path - why not cache the unpacked key?

**Comment by yosefe** [comment]
the goal of wrapping it with pack/unpack is to create abstraction for some transports which actually do stuff in the unpack() such as mapping shared memory

**Comment by hjelmn** [comment]
Ah, ok. That makes sense. For a btl I only care about ib (and ugni somewhat). I will have to look at what they do in unpack.

**Comment by hjelmn** [comment]
Caching is a little harder in the btl interface than in other contexts. There is no explicit call to turn a registration into a remote registration. UCT chose to abstract the rkey so that leads to this mismatch.

**Comment by hjelmn** [comment]
Looks like the overhead really is just the malloc/free when it comes to ugni.

**Comment by hjelmn** [comment]
ib doesn't have this problem :+1:

**Comment by hjelmn** [comment]
i think i can fix it easily for ugni.

**Comment by hjelmn** [comment]
Ok, yeah. Fixed it. PR incoming.

**Comment by hjelmn** [comment]
Ugh, or not. The buffer passed in could go away couldn't it. Will see what I can do.

**Comment by shamisp** [comment]
Agree with @yosefe here. it should not be called in critical path.

**Comment by hjelmn** [comment]
Really difficult to avoid in the context of a btl though. I can create an rcache but it isn't needed on ib as the unpack is almost free.

The unpack performance probably isn't a big issue in the long run though. The performance impact is only really going to be on benchmarks and will not be seen by apps.

**Comment by shamisp** [comment]
@hjelmn I'm not sure why it is in critical path, and why it is so expensive with ugni. Most likely I'm missing something. 
Is this because of malloc call here https://github.com/openucx/ucx/blob/master/src/uct/ugni/base/ugni_md.c#L146

Can we avoid it by using memory pool within unpack ?

**Comment by hjelmn** [comment]
Its in the critical path because the btl interface doesn't have the concept of pack/unpack. The key returned by register is the remote key. As such that means that each call into the btl needs to unpack the key because it doesn't know whether the registration has been seen before, if something has changed, etc. It might be worth adding pack/unpack to the btl but that probably won't happen in the v4.0.0 timeframe.

For now if you can remove the malloc that should greatly speed things up. It will bring the difference in very small messages for single threaded puts with osc/rdma + btl/ucx much closer to the performance of osc/ucx on ugni. The rest of the difference is overhead in osc/rdma I am trying to figure out.

**Comment by shamisp** [comment]
I can craft a patch (it is relatively small), but I don't have ugni headers to compile it. I believe it should be enough to convey the idea how it can be fixed.

**Comment by shamisp** [comment]
@MattBBaker do you have cycles for this ? We need to initialize and a memory pool for this ?

**Comment by MattBBaker** [comment]
The mpool maybe the best short term solution. It will be a little bit before I have the free cycles to hack this up, but it shouldn't be too hard.

**Comment by shamisp** [comment]
what is the long term ?

**Comment by yosefe** [comment]
I think we can finalize it in 1-2 years, now it’s too early
there are many things we plan to add to UCT (e.g datatypes)

**Comment by shamisp** [comment]
I think, it is a bit unrelated to the discussion...

**Comment by MattBBaker** [comment]
I'd like to think through a solution to not allocate any memory. Though the mpool plan may become the long term solution.

**Comment by hjelmn** [comment]
Only thing I can think of is may uct_rkey_t a 128-bit type.

**Comment by hjelmn** [comment]
Actually, that looks like a fairly simple change. Let me open a PR and see what people think.

**Comment by yosefe** [comment]
larger rkey will add overhead to other transports

**Comment by hjelmn** [comment]
would it really? if you are worried about memory, how many rkeys do you expect to be in use at any time? if you are worried about performance, it would be easy to test and see if there is any hit (i really doubt it.)

**Comment by hjelmn** [comment]
i agree that if it adds overhead for the ib or sm it doesn't make sense

**Comment by yosefe** [comment]
since there are no real 128-bit types, it would add memory loads/stores instead of being passed around in a register (or you suggest to use sse types)?
i think that even if the overhead is not measurable in a specific test/setup it's still not worth it, mpool allocation is a better solution.

**Comment by hjelmn** [comment]
Ok, btls don't suffer at all due to the large max key size but pointers to keys are what is passed around.

**Comment by hjelmn** [comment]
fwiw, the performance on ugni with a struct rkey is pretty good:

```
mpirun  --mca osc ^ucx --mca btl_uct_transports ugni-ugni_rdma --mca osc_rdma_btls uct,openib --mca osc_base_verbose 0 -n 2 -N 1 --bind-to socket /lustre/ttscratch1/hjelmn/rmamt_benchmarks/src/rmamt_bw -x -t 32 -o put -s flush -i 5000
##########################################
# RMA-MT Bandwidth
#
# Operation: put
# Sync: flush
# Thread count: 32
# Iterations: 5000
# Ibarrier: no, sleep interval: 10000ns
# Bind worker threads: yes
# Number of windows: 1
##########################################
  BpT(32)	  BxT(32)	Bandwidth(MiB/s)	Message_Rate(M/s)
        1	       32	       24.602		 25797031.536
        2	       64	       47.954		 25141562.712
        4	      128	       91.079		 23875846.790
        8	      256	      201.943		 26469118.810
       16	      512	      387.117		 25370094.177
       32	     1024	      801.040		 26248463.029
       64	     2048	     1525.516		 24994060.005
      128	     4096	     2916.770		 23894177.467
      256	     8192	     4634.684		 18983666.690
      512	    16384	     6273.043		 12847192.363
     1024	    32768	     7659.769		  7843603.257
     2048	    65536	     8178.057		  4187165.020
     4096	   131072	     4968.855		  1272026.895
     8192	   262144	     5532.505		   708160.681
    16384	   524288	     6210.715		   397485.791
    32768	  1048576	     6898.626		   220756.049
    65536	  2097152	     7416.259		   118660.143
   131072	  4194304	     7718.863		    61750.906
```

**Comment by hjelmn** [comment]
(ignore) the large message performance, that is a bug in uct/ugni.

**Comment by hjelmn** [comment]
This is within 10% of btl/ugni. Without the change it struggles to get 50% at 32 threads.

**Comment by hjelmn** [comment]
Problem is there is nothing to cache on. There is no interface to say an rkey is associated with an address range and endpoint. It would take another api to make that work. i don't think that makes any sense. @shamisp what do you think?

**Comment by hjelmn** [comment]
Another data point. It can be up to within 10%:
```
  BpT(32)	  BxT(32)	Bandwidth(MiB/s)	Message_Rate(M/s)
        1	       32	       29.456		 30886719.097
        2	       64	       58.822		 30839794.592
        4	      128	       86.987		 22803229.507
        8	      256	      214.649		 28134504.737
       16	      512	      459.182		 30092983.558
       32	     1024	      909.646		 29807292.130
       64	     2048	     1634.671		 26782447.186
      128	     4096	     3002.218		 24594165.526
      256	     8192	     4691.925		 19218124.998
      512	    16384	     6389.403		 13085497.286
     1024	    32768	     7786.263		  7973133.729
```

btl/ugni:
```
  BpT(32)	  BxT(32)	Bandwidth(MiB/s)	Message_Rate(M/s)
        1	       32	       32.412		 33986157.863
        2	       64	       65.338		 34256004.542
        4	      128	      141.459		 37082540.404
        8	      256	      284.495		 37289318.265
       16	      512	      525.955		 34468978.458
       32	     1024	     1050.232		 34413989.975
       64	     2048	     1813.152		 29706685.471
      128	     4096	     3172.141		 25986175.355
      256	     8192	     4779.911		 19578516.152
      512	    16384	     6503.927		 13320043.260
     1024	    32768	     7775.611		  7962225.609
```

**Comment by hjelmn** [comment]
Also, on x86 it there aren't even extra instructions:

```c
ucs_status_t uct_md_stub_rkey_unpack(uct_md_component_t *mdc,
                                     const void *rkey_buffer, uct_rkey_t *rkey_p,
                                     void **handle_p)
{
    rkey_p->key[0]   = 0xdeadbeef;
    rkey_p->key[1]   = 0xfeedbeef;
    *handle_p = NULL;
    return UCS_OK;
}
```

```
00000000000005e0 <uct_md_stub_rkey_unpack>:
 5e0:   66 0f 6f 05 00 00 00    movdqa 0x0(%rip),%xmm0        # 5e8 <uct_md_stub_rkey_unpack+0x8>
 5e7:   00 
 5e8:   31 c0                   xor    %eax,%eax
 5ea:   0f 29 02                movaps %xmm0,(%rdx)
 5ed:   48 c7 01 00 00 00 00    movq   $0x0,(%rcx)
 5f4:   c3                      retq   
 5f5:   90                      nop
 5f6:   66 2e 0f 1f 84 00 00    nopw   %cs:0x0(%rax,%rax,1)
 5fd:   00 00 00 
```

**Comment by yosefe** [comment]
what about the code in UCP which handles rkeys?

**Comment by hjelmn** [comment]
Will check. I suspect it uses SSE there as well.

**Comment by yosefe** [comment]
BTW it will also increase UCP rkey structure

**Comment by hjelmn** [comment]
yup. i will make sure i check everywhere to see what the actual effect on instruction count is. though instruction counting is really not the best metric as any extra instructions might end up in otherwise empty pipeline slots.

**Comment by shamisp** [comment]
@hjelmn - agreed - instruction count is the next to useless metric :)

**Comment by shamisp** [comment]
On armv8 has neon instructions that in theory supposed to handle 128B ops.

**Comment by hjelmn** [comment]
@shamisp Did you intend to close this?

**Comment by shamisp** [comment]
Nope !

**Comment by hjelmn** [comment]
Yeah, they need to make comment and close ask before it closes.
